### ISSUE TITLE
Stable identifiers for analysis errors/warnings/hints/lints.

### ISSUE THREAD

**Issue by pq** [dart-lang/sdk#24522]
A prerequisite to the proposed means of suppressing unwanted warnings (https://github.com/dart-lang/sdk/issues/12490) and relatedly make others fatal (https://github.com/dart-lang/sdk/issues/24452) is a way to uniquely identify the warnings to suppress.

In other words we want to be able to do something like:

```
analyzer:
  warnings:
    exclude:
     - UNIQUE_ID
```

or:

```
// [analyzer]: ignore: UNIQUE_ID
var unused = "Unused but I'm OK with that.";
```

This issue tracks the effort to specify those `UNIQUE_ID`s.

**Comment by pq** [comment]
(Paging @bwilkerson @gbracha @skybrian for updates on their related efforts.)

**Comment by bwilkerson** [comment]
Gilad is on vacation, but he and I are working to define the errors and warnings that should have unique identifiers. After those identifiers are defined we can work on getting that information into analyzer so that it's usable for that purpose.

**Comment by skybrian** [comment]
It's also important to define a format that allows non-standard ids for linters, etc. I'd like to see a two-level namespace similar to mime types. So, ids that are part of the Dart standard might begin with "dart/".

Other ids could use a Dart package name followed by "/" as their namespace prefix. We should probably define allowable characters.

**Comment by sethladd** [comment]
This is a pretty broad issue (for all errors, warnings, hints, and lints). Should we create separate issues?

**Comment by bwilkerson** [comment]
I think we need a consistent scheme for providing these identifiers, so it's probably best to keep a single issue for now.

**Comment by sethladd** [comment]
This probably blocks https://github.com/dart-lang/sdk/issues/24680 then, so moving this to 1.14.

**Comment by mit-mit** [comment]
@bwilkerson, any news on this item? Are we on track for getting this done in 1.14?

**Comment by lrhn** [comment]
@floitschG

**Comment by bwilkerson** [comment]
@lrhn @floitschG Note that this is *not* the same issue as Florian's proposal to have shared messages. In the analyzer, in order to improve feedback, we often produce multiple different messages for a single error. This issue is about creating unique identifiers for those errors so that all of the messages can be disabled simultaneously.

@mit-mit No, Given the limited amount of time before 1.14 (partially as a result of the holidays), I do not think that this will happen by then. I'll move it to 1.15.

**Comment by skybrian** [comment]
@bwilkerson "we often produce multiple different messages for a single error. This issue is about creating unique identifiers for those errors so that all of the messages can be disabled simultaneously."

I'm having trouble understanding this. Do you want to split up error codes (use multiple codes where we currently use a single code) or merge them (use a single code where we currently have multiple codes)?

**Comment by bwilkerson** [comment]
I draw a distinction between the error and the message produced in order to report the error. For example, the specification defines the following error:

> Let C be a concrete class that does not have a noSuchMethod() method distinct from the one declared in class Object. It is a static warning if the implicit interface of C includes an instance member m of type F and C does not declare or inherit a corresponding non-abstract instance member m of type
F' such that F' <: F.

We produce different messages depending on how many members there are that C is missing. If there are few enough we just list them, otherwise we list the first few and indicate that there are N more.

Hence, one error produces multiple messages.

Florian has a proposal to associate unique identifiers with each message. His purpose, which I support, is to share messages across our tools in order to reduce confusion for users.

I think we also need to have unique identifiers for each error so that users can disable errors. Users want to disable errors, not messages, and shouldn't be required  to enumerate all of the messages associated with an error, nor to update their preferences to include a new message if next week we decide to add another variant of a message for the same error.

**Comment by skybrian** [comment]
Okay, so to continue this example, in analyzer/lib/src/generated/error.dart, the codes currently look like this:

NON_ABSTRACT_CLASS_INHERITS_ABSTRACT_MEMBER_ONE
NON_ABSTRACT_CLASS_INHERITS_ABSTRACT_MEMBER_TWO
NON_ABSTRACT_CLASS_INHERITS_ABSTRACT_MEMBER_THREE
NON_ABSTRACT_CLASS_INHERITS_ABSTRACT_MEMBER_FOUR
NON_ABSTRACT_CLASS_INHERITS_ABSTRACT_MEMBER_FIVE_PLUS

I agree that those codes need to be merged before users start relying on them. (Before this I've only seen codes that don't have this problem.)

**Comment by bwilkerson** [comment]
> I agree that those codes need to be merged before users start relying on them.

Not sure whether we're in agreement or not. If you mean that you agree that there can be a one-to-many relationship between the errors that users control and the messages that tools display, and that we need unique identifiers for errors, then we're in agreement. If you mean that we need to collapse these messages into a single message so that there's a one-to-one mapping, then I disagree,

**Comment by skybrian** [comment]
Terminology is tricky, but I think we agree.

It doesn't help that the analyzer uses a "StaticWarningCode" class to represent a (code, message) pair. I think it's fine to report multiple (slightly different) messages for the same code.

**Comment by bwilkerson** [comment]
> It doesn't help that the analyzer uses a "StaticWarningCode" class to represent a (code, message) pair.

:-) I didn't want to get into it, but yes, that's a third distinct kind of code (error codes, message codes, and implementation-level codes?). One that might be able to go away after we have the other two kinds in place.

**Comment by zoechi** [comment]
I think these error names should be printed with the error message otherwise it's often hard to infer the error code from the printed message. Above names are IMHO too long to be printed with the error message.
I think there should be a short error code associated with the error name and this code should then be printed with the error message.
Ideally the short error codes should be in a form that makes them searchable. Searching for `dart 123` in Google won't lead to meaningful results on the first 10 pages. I guess `dart err_1r2h3` would make false positives less likely but probably still room for improvement. Something like an UID but shorter like http://wiki.freebase.com/wiki/Machine_ID

**Comment by bwilkerson** [comment]
I agree; they should be short and searchable. The existing error "codes" were never intended for external consumption. They are just the names of the class fields in our original Java implementation of the analyzer. We can do much better.

**Comment by skybrian** [comment]
I think the current codes are well-named and searchable, but unfortunately they are quite long. If we want them to be short, I think there's little choice but to use error numbers or similar codes, which won't be readable but can at least be made searchable and unambiguous. Perhaps something like Oracle's error-numbering scheme [1] with a prefix for each group of messages.

[1] http://ss64.com/ora/syntax-errors.html

**Comment by mit-mit** [comment]
Clearing out 1.15 milestone as the last full push to dev has happened. If any changes are required before 1.15 is shipped, please file a merge request

**Comment by floitschG** [comment]
I think this is mostly the work I'm doing.
I'm reassigning to me, and clear the milestone.
At some point I will probably assign back to Brian.
